### ISSUE TITLE
The symbol numbers in the dot-num files need to be finalized

### ISSUE THREAD

**Issue by mspncp** [openssl/openssl#12868]
While working on #12866, it occurred to me that the libcrypto.num file currently contains a lot of question mark placeholders.

https://github.com/openssl/openssl/blob/08073700cc50bcd0df5c0ee68c100e300a320d03/util/libcrypto.num#L4425-L4429

This feature was added to `make update` in order to reduce the likelyhood of merge conflicts in the dot-num files. If I recall correctly, we decided in the last F2F to implement the following:

1. Add question mark placeholders instead of actual numbers for newly added functions.
2. Add new entries in lexicographical order instead of adding them to the end.
3. Finalize the placeholders on release, i.e., replace them by actual numbers.

It seems to me that only part 1) has been implemented, but I might have missed something. If it's not complete yet, then **this feature must be completed before the beta release**, for following reasons:

- Part 3) is absolutely necessary to guarantee API stability.
- Part 2) is the crucial part for reducing the probability of merge collisions, together with part 1.

If part 3) hat not been implemented yet, I would suggest to add a, `make release_update` or (`make final_update`) target for doing it manually. Alternatively, one could do it automatically in the release script, unless it's an alpha release.

**Note:** Assuming that all three points are implemented before the beta1 release, we could optionally renumber all symbols such that the dot-num files are initially sorted alphabetically _and_ numerically in 3.0.0 beta1. This could be easily done by replacing all existing numbers by question marks and running `make release_update` once.

_(I would have labelled this issue as [triaged: technical debt] if that label had already existed.)_

**Comment by richsalz** [comment]
I don't believe the project is guaranteeing that the BETA will be binary-compatible with the FINAL release.  But, rather, that there will be no API's from now on.

**Comment by mspncp** [comment]
Ok, in that case it would only be a todo for the final release.

**Comment by mspncp** [comment]
I changed the title accordingly. @levitte @mattcaswell if you think this issue does not need to be resolved for the beta1 release, please feel free to move it back to the 3.0.0 milestone.

**Comment by mspncp** [comment]
> I don't believe the project is guaranteeing that the BETA will be binary-compatible with the FINAL release. But, rather, that there will be no API's from now on.

Can you confirm this, @levitte @mattcaswell?

**Comment by paulidale** [comment]
This is my understanding.  BETA is API not ABI.

**Comment by mspncp** [comment]
Ok, that means it can be postponed to the 3.0.0 milestone.

**Comment by levitte** [comment]
> in order to reduce the likelyhood of merge conflicts in the dot-num files

That isn't _quite_ true...  The intent was to reduce the churn of having to renumber.  These days, a conflict is much more easily resolved by just removing the conflict markers (not always, there are times, although fairly rare, where a bit more editing is required, but all the same, not having to renumber was a huge relief)

**Comment by levitte** [comment]
> I don't believe the project is guaranteeing that the BETA will be binary-compatible with the FINAL release. But, rather, that there will be no API's from now on.

I share this view

**Comment by levitte** [comment]
> 2.  Add new entries in lexicographical order instead of adding them to the end.

Er, did we actually *decide* this?  My recollection is that it could be a nice to have, but wasn't absolutely necessary, and I believe we concluded that it's no clear that it would reduce merge conflicts...

**Comment by levitte** [comment]
> 3.  Finalize the placeholders on release, i.e., replace them by actual numbers.

FYI, this is implemented, and much more automated than you might think.

https://github.com/openssl/openssl/blob/ecf15b16ee8223a9a383b97ee41126fbedf89bb5/util/mknum.pl#L91-L96

**Comment by paulidale** [comment]
This is done automatically by the release script that will produce the beta.

**Comment by mspncp** [comment]
Ok, I'll close the ticket.
